tuf/api/metadata.py

```diff
@@ -127,7 +124,8 @@ def from_dict(cls, metadata: Dict[str, Any]) -> "Metadata":
             signatures[sig.keyid] = sig
 
         return cls(
-            signed=inner_cls.from_dict(metadata.pop("signed")),
+            # Specific type T is not known at static type check time: cast
+            signed=cast(T, inner_cls.from_dict(metadata.pop("signed"))),
```

Comment:
well, it's useful to remember that `T` and `cast()` have zero runtime effect: cast() is me promising to mypy that signed will be one of the types T once the code runs, and telling mypy to not worry about it.

For the static typing to work `__init__() / from_dict()` do not need to define what type T is: calling methods like `Root.metadata_from_file()` do define it (as the return value is `Metadata[Root]`).

Reply:
You can see this in action if you use the current constructor with this branch: `Metadata.from_file()` still works but return value is not statically typed with the correct type (it's just Metadata[T]) because there is nothing the static type check could use to figure it out.

conveniently at least VS Code offers me all the possible completions (Root,Targets,Snapshot,Timestamp) when T is not defined so that's a win-win.